osbs/tekton.py

```diff
@@ -0,0 +1,709 @@
+                    j = json.loads(line.decode(encoding))
+                except ValueError:
+                    logger.error("Cannot decode watch event: %s", line)
+                    continue
+
+                if 'object' not in j:
+                    logger.error("Watch event has no 'object': %s", j)
+                    continue
+
+                if 'type' not in j:
+                    logger.error("Watch event has no 'type': %s", j)
+                    continue
+
+                yield (j['type'].lower(), j['object'])
+
+            log_and_sleep()
+
+
+class PipelineRun():
+    def __init__(self, os, pipeline_name, pipeline_run_name):
+        self.os = os
+        self.pipeline_name = pipeline_name
+        self.pipeline_run_name = pipeline_run_name
+        self.api_path = 'apis'
+        self.api_version = API_VERSION
+
+    def start_pipeline_run(self):
+        url = self.os._build_url(
+            self.api_path,
+            self.api_version,
+            "pipelineruns"
+        )
+        data = {
+            "apiVersion": API_VERSION,
+            "kind": "PipelineRun",
+            "metadata": {
+                "name": self.pipeline_run_name
+            },
+            "spec": {
+                "pipelineRef": {
+                    "name": self.pipeline_name
+                }
+            }
+        }
+        return self.os._post(url, data=json.dumps(data),
+                          headers={"Content-Type": "application/json", "Accept": "application/json"})
+
+    def cancel_pipeline_run(self):
+        url = self.os._build_url(
+            self.api_path,
+            self.api_version,
+            f"pipelineruns/{self.pipeline_run_name}"
+        )
+        data = {
+            "apiVersion": API_VERSION,
+            "kind":"PipelineRun",
+            "metadata":{
+                "name":self.pipeline_run_name,
+                "namespace": self.namespace
+            },
+            "spec":{
+                "pipelineRef":{
+                    "name":self.pipeline_name
+                },
+                "status": "PipelineRunCancelled"
+            }
+        }
+        return self.os._patch(url, data=json.dumps(data),
+                          headers={"Content-Type": "application/merge-patch+json", "Accept": "application/json"})
+
+    def update_labels(self, labels):
+        url = self.os._build_url(
+            self.api_path,
+            self.api_version,
+            f"pipelineruns/{self.pipeline_run_name}"
+        )
+        data = {
+            "apiVersion": API_VERSION,
+            "kind":"PipelineRun",
+            "metadata": {
+                "name": self.pipeline_run_name,
+                "namespace": self.namespace,
+                "labels": labels
+            },
+            "spec": {
+                "pipelineRef": {
+                    "name":self.pipeline_name
+                }
+            }
+        }
+        return self.os._patch(url, data=json.dumps(data),
```

Comment:
Will this update or rewrite all labels?

Original implementation updates labels, i.e. it first downloads current labels